src/contexts/providers.context.tsx

```diff
@@ -18,7 +18,7 @@ import routes from "src/routes";
 import { Chain, EthereumEvent, WalletName } from "src/domain";
 
 interface ProvidersContext {
-  connectedProvider?: Web3Provider;
+  connectedProvider?: { provider: Web3Provider; chainId: number };
   account: AsyncTask<string, string>;
   isConnectedProviderChainOk: (chain: Chain) => Promise<boolean>;
```

Comment:
This function should not return a promise anymore.
Indeed, now that the chainId is syncronously available, this function should not exist at all. We can replace the calls to this function with:

```typescript
if (connectedProvider && connectedProvider.chainId === chainId) {...}
```

Reply:
This should not be like this. I see 2 problems in the current implementation:

* The signature of `changeNetwork` should be: `(chain: Chain) => Promise<number>` and its implementation should change the network and then return the `chainId` of the newly selected network.
* After the network has been changed, we need to call `setConnectedProvider` and update the value of `chainId` to the new selected network.

Sadly, I see that MetaMask resolves to `null` in both `wallet_switchEthereumChain` and `wallet_addEthereumChain` after changing the `Network`, so our implementation should retrieve the network after the change to be able to return it and store it in the `connectedProvider` state.

I'd also suggest to split the function `changeNetwork` into 3 functions like this to have a clear separation of concerns:

```typescript
const switchNetwork = (chain: Chain, connectedProvider: Web3Provider): Promise<number> => {
  if (connectedProvider.provider.request) {
    return connectedProvider.provider
      .request({
        method: "wallet_switchEthereumChain",
        params: [{ chainId: hexValue(chain.chainId) }],
      })
      .then(async () => {
        const { chainId } = await connectedProvider.getNetwork(); // we retrieve the current network
        setConnectedProvider({ provider: connectedProvider, chainId: chainId }); // update our state
        return Promise.resolve(chainId); // and return the new value
      });
  }
  return Promise.reject(new Error("The provider does not have a request method"));
};

const addAndSwitchNetwork = (chain: Chain, connectedProvider: Web3Provider): Promise<number> => {
  if (connectedProvider.provider.request) {
    return connectedProvider.provider
      .request({
        method: "wallet_addEthereumChain",
        params: [
          {
            chainId: hexValue(chain.chainId),
            chainName: getChainName(chain),
            rpcUrls: [chain.provider.connection.url],
          },
        ],
      })
      .then(async () => {
        const { chainId } = await connectedProvider.getNetwork(); // we retrieve the current network
        setConnectedProvider({ provider: connectedProvider, chainId: chainId }); // update our state
        return Promise.resolve(chainId); // and return the new value
      });
  }
  return Promise.reject(new Error("The provider does not have a request method"));
};

const changeNetwork = useCallback(
  (chain: Chain) => {
    if (connectedProvider && connectedProvider.provider.provider.isMetaMask) {
      return switchNetwork(chain, connectedProvider.provider).catch((error) => {
        if (isMetamaskUnknownChainError(error)) {
          return addAndSwitchNetwork(chain, connectedProvider.provider);
        } else {
          throw error;
        }
      });
    } else {
      return Promise.reject(new Error(providersContextNotReadyErrorMsg));
    }
  },
  [connectedProvider]
);
```